pkg/plugin/standalone/registry.go

```diff
@@ -15,28 +15,197 @@
+	}
+
+	if pluginDir != "" {
+		// extract absolute path to make it clearer in the logs what directory is used
+		absPluginDir, err := filepath.Abs(pluginDir)
+		if err != nil {
+			r.logger.Warn(context.Background()).Err(err).Msg("could not extract absolute plugins path")
+		} else {
+			r.pluginDir = absPluginDir // store plugin dir for hot reloads
+			r.reloadPlugins()
+		}
+	}
+
+	r.logger.Info(context.Background()).
+		Str(log.PluginPathField, r.pluginDir).
+		Int("count", len(r.List())).
+		Msg("standalone plugins initialized")
+
+	return r
+}
+
+func newFullName(pluginName, pluginVersion string) plugin.FullName {
+	return plugin.NewFullName(plugin.PluginTypeStandalone, pluginName, pluginVersion)
+}
+
+func (r *Registry) reloadPlugins() {
+	plugins := r.loadPlugins(context.Background(), r.pluginDir)
+	r.m.Lock()
+	r.plugins = plugins
+	r.m.Unlock()
+}
+
+func (r *Registry) loadPlugins(ctx context.Context, pluginDir string) map[string]map[string]blueprint {
+	r.logger.Debug(ctx).Msgf("loading plugins from directory %v", pluginDir)
+	plugins := make(map[string]map[string]blueprint)
+
+	dirEntries, err := os.ReadDir(pluginDir)
+	if err != nil {
+		r.logger.Warn(ctx).Err(err).Msg("could not read plugin directory")
+		return plugins // return empty map
+	}
+	warn := func(ctx context.Context, err error, pluginPath string) {
+		r.logger.Warn(ctx).
+			Err(err).
+			Str(log.PluginPathField, pluginPath).
+			Msgf("could not load standalone plugin")
+	}
+
+	for _, dirEntry := range dirEntries {
+		if dirEntry.IsDir() {
+			// skip directories
+			continue
+		}
+
+		pluginPath := path.Join(pluginDir, dirEntry.Name())
+
+		// create dispenser without a logger to not spam logs on refresh
+		dispenser, err := standalonev1.NewDispenser(zerolog.Nop(), pluginPath)
+		if err != nil {
+			err = cerrors.Errorf("failed to create dispenser: %w", err)
+			warn(ctx, err, pluginPath)
+			continue
+		}
+
+		specPlugin, err := dispenser.DispenseSpecifier()
+		if err != nil {
+			err = cerrors.Errorf("failed to dispense specifier (tip: check if the file is a valid plugin binary and if you have permissions for running it): %w", err)
+			warn(ctx, err, pluginPath)
+			continue
+		}
+
+		specs, err := specPlugin.Specify()
+		if err != nil {
+			err = cerrors.Errorf("failed to get specs: %w", err)
+			warn(ctx, err, pluginPath)
+			continue
+		}
+
+		versionMap := plugins[specs.Name]
+		if versionMap == nil {
+			versionMap = make(map[string]blueprint)
+			plugins[specs.Name] = versionMap
+		}
+
+		fullName := newFullName(specs.Name, specs.Version)
+		if conflict, ok := versionMap[specs.Version]; ok {
+			err = cerrors.Errorf("conflict detected, plugin %v already registered, please remove either %v or %v, these plugins won't be usable until that happens", fullName, conflict.path, pluginPath)
+			warn(ctx, err, pluginPath)
+			// delete plugin from map at the end so that further duplicates can
+			// still be found
+			defer func() {
```

Comment:
We need to leave the conflicting version in the map until we loop through all plugins in case we find more duplicates. Only after we have gone through all of them can we remove the original conflicting plugin, so that none of that conflicting plugin are left.

If we remove it right away and we have 3 plugins that report the same name and version, we will be left with 1 plugin in the registry after we load them.

Reply:
Gotcha. For some dumb reason, when reading delete(map, specs.Version), I somehow understood it's only the "new" duplicate being removed, which makes no sense.